### PR TITLE
chore(debugger): Docs

### PR DESCRIPTION
**Pull request by mverzilli**
# Description

Adds documentation for the Noir debugger.

## Problem

Part of #3015.

## Summary

Adds quickstart, how to's and reference pages to Noir's docsite covering both the VS Code and REPL debuggers.

## Documentation

Check one:
- [ ] No documentation needed.
- [x] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by signorecello** [comment]
dammit seems like it doesn't add the documentation label when non-maintainers add stuff to docs, so it doesn't generate a preview... bah, need to fix that because it's annoying

File: docs/docs/how_to/debugger/debugging_with_the_repl.md
Comment (signorecello) [inline docs/docs/how_to/debugger/debugging_with_the_repl.md:93]:
```suggestion
Some commands operate only for unconstrained functions, such as `memory`, `memset`, `registers`, `regset`. If you try to use them while execution is paused at an ACIR opcode, the debugger will simply inform you that you are not executing unconstrained code:
```
As much as I like the naming, I believe we're dropping "Brillig" in favour of the more self-explanatory name "unconstrained"

File: docs/docs/how_to/debugger/debugging_with_vs_code.md
Comment (signorecello) [inline docs/docs/how_to/debugger/debugging_with_vs_code.md:47]: Same for the Brilling naming, just keeping in mind (I won't flag this every time I see it moving further on this review)

File: docs/docs/noir/tooling/debugger.md
Comment (signorecello) [inline docs/docs/noir/tooling/debugger.md:36]: Yeah I wouldn't put a quickstart here, feel free to simply mention that Nargo ships with the `debug` command (point to reference) and link to the how-to-guide.

Reason is mostly to be aligned with our docs framework Diátaxis and reducing maintainability in when changes come to the debugger

File: docs/docs/noir/tooling/_category_.json
Comment (signorecello) [inline docs/docs/noir/tooling/_category_.json:1]: Well the decision to have "tooling" inside `getting_started` was fairly deliberate, as it doesn't necessarily feed into the "concepts" section which is more about explanations and reference. Plus, it is expected that when getting started, devs will be wanting to set up some tooling including the debugger and the vs extension.

The only other place that we've considered is right at the top-level. But we didn't (and still don't) have enough tooling to justify its own menu.

We're open to suggestions though if you believe the tooling section would be best placed inside "noir" section

File: docs/docs/reference/debugger/debugger_repl.md
Comment (signorecello) [inline docs/docs/reference/debugger/debugger_repl.md:265]:
```suggestion
:::note

The ability to inspect variable values from the debugger depends on compilation to be run in a special debug instrumentation mode. This instrumentation weaves variable tracing code with the original source code. 

:::
```

Using admonitions block is highly encouraged

File: docs/docs/reference/debugger/debugger_repl.md
Comment (signorecello) [inline docs/docs/reference/debugger/debugger_repl.md:388]: Same for admonition blocks for the rest of the page.

Another thing to note: we are planning to have reference documentation be auto-generated with CLAP. This is on my TODO list. Would be much appreciated, if you're comfortable with CLAP, to add these docs straight into the debugger code. Otherwise just leave it and I'll find my way somehow 😄

**Review by signorecello — CHANGES_REQUESTED** [review]
Some quick requested changes. Great work so far!

File: docs/docs/noir/tooling/_category_.json
Comment (mverzilli) [reply docs/docs/noir/tooling/_category_.json:1]: Oh I thought you meant that when we first discussed the docs structure: "so on top of my mind, I'd do three things:
would move the "Tooling" into the Noir section:"

I agree that it's not a perfect match under the Noir Language section.

I think I would expect to find it at the root level, say the last menu as a sibling of `Getting Started` and `The Noir Language`.  Or as the next menu after `Reference`. I think the fact that there are few tools doesn't change that.

That said, if you'd prefer things to stay at getting started that works too! :)

**Comment by signorecello** [comment]
> > dammit seems like it doesn't add the documentation label when non-maintainers add stuff to docs, so it doesn't generate a preview... bah, need to fix that because it's annoying
> 
> Maybe it's because I opened it as draft?

Looking at the GH action, I think the labeling problem is just a security feature. Maybe the token used doesn't have those permissions.

As for the rest of the CI run, I don't know. The "deploy PR" action didn't seem to run even.

File: docs/docs/noir/tooling/_category_.json
Comment (mverzilli) [reply docs/docs/noir/tooling/_category_.json:1]: 🙌 on it

File: docs/docs/how_to/debugger/debugging_with_the_repl.md
Comment (mverzilli) [reply docs/docs/how_to/debugger/debugging_with_the_repl.md:93]: I'll drop it from the docs explanations, but what's your opinion on the command descriptions show by the REPL itself?

For example:

`memory                           show Brillig memory (valid when executing a Brillig block)`

Some commands of the debugger are exposing the BrilligVM's internals, so the vocabulary for those will be kind of tricky to get right (it's one thing to talk about unconstrained code when explaining from Noir's perspective, but the semantic category of things like memory or registers doesn't match).

File: docs/docs/how_to/debugger/debugging_with_vs_code.md
Comment (mverzilli) [reply docs/docs/how_to/debugger/debugging_with_vs_code.md:47]: Same comment as above, I see how to use `unconstrained` instead Brillig when talking concepts, but for VM components such as registers it becomes a bit trickier. 

What would you call "Brillig registers"? Unconstrained VM registers? Unconstrained runtime registers?

File: docs/docs/reference/debugger/debugger_repl.md
Comment (mverzilli) [reply docs/docs/reference/debugger/debugger_repl.md]: Done!

File: docs/docs/noir/tooling/debugger.md
Comment (mverzilli) [reply docs/docs/noir/tooling/debugger.md]: Done!

File: docs/docs/noir/tooling/_category_.json
Comment (mverzilli) [reply docs/docs/noir/tooling/_category_.json:1]: Moved Tooling to root level, under "Reference". I couldn't find a way to make a root level menu also have its own content (as it used to be the case with Tooling under Getting Started), so renamed the old tooling index to "other tools", and now it's just another page under Tooling. Let me know if this doesn't work.

File: docs/docs/reference/debugger/debugger_repl.md
Comment (mverzilli) [reply docs/docs/reference/debugger/debugger_repl.md]: Re: CLAP, will address on a separate PR

**Comment by mverzilli** [comment]
Thanks for the review @signorecello! Addressed most of the comments, except: 

- The part about adding more of these docs to CLAP (will work on that on a different PR).

Regarding Brillig=>unconstrained language, I replaced it in as many places as it felt adequate, but it didn't seem right to rename things like "Brillig registers" or "Brillig memory". I ultimately defer to your decision on that, though.

File: docs/docs/how_to/debugger/debugging_with_vs_code.md
Comment (signorecello) [reply docs/docs/how_to/debugger/debugging_with_vs_code.md:47]: Unconstrained VM registers looks ok... @Savio-Sou any thoughts?

**Comment by signorecello** [comment]
Yes we really want to avoid using the word Brillig. Not that we don't like Lewis Carroll, but because it would involve yet another definition, specially one with a creative yet quite non-descriptive name

File: docs/docs/how_to/debugger/debugging_with_the_repl.md
Comment (michaeljklein) [inline docs/docs/how_to/debugger/debugging_with_the_repl.md:138]: What does it mean to "hydrate" `x`?

Comment (mverzilli) [reply docs/docs/how_to/debugger/debugging_with_the_repl.md:138]: I meant that it's now tracked by debugger, but I definitely see it's a bit of an arcane word choice. Will try to simplify :)

Comment (mverzilli) [reply docs/docs/how_to/debugger/debugging_with_the_repl.md]: Hopefully that's a bit clearer?

https://github.com/noir-lang/noir/pull/4145/commits/147b7d7d680c77d0315bf7e4410fce5a61a5d99e#diff-bcfff321a280415fea2e80447b8c878910193e6040a547b6ab479c1a161a0abeR138

**Review by signorecello — APPROVED** [review]
LGTM, thanks!

**Comment by Savio-Sou** [comment]
Sorry for the delay and thanks for the great work here!

**Pull request merged**
